### PR TITLE
base64 encoding decoding fixes to pre/postsubmit luci_subscription ha…

### PR DESCRIPTION
**Pull request by yusuf-goog**
…ndlers.

Some pubsub messages seem to be base64 decoding/encoding differently than others. Adding an
additional decode if we encounter a FormatException.

Bug: flutter/flutter#107961

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by keyonghan** [comment]
This is another place using userData: https://cs.opensource.google/flutter/cocoon/+/main:app_dart/lib/src/service/github_checks_service.dart;l=74, and we need to fix it as well.

Let's create a util function to parse the string so that all those four locations can simply call it.

**Comment by yusuf-goog** [comment]
> This is another place using userData: https://cs.opensource.google/flutter/cocoon/+/main:app_dart/lib/src/service/github_checks_service.dart;l=74, and we need to fix it as well.
> 
> Let's create a util function to parse the string so that all those four locations can simply call it.

This is more a patch for to fix the outage event we are having. We should figure out why we even have these differently encoded messages, and make sure it doesn't happen again.

File: app_dart/lib/src/request_handlers/scheduler/request_subscription.dart [hunk 1]
```diff
@@ -47,7 +47,12 @@ class SchedulerRequestSubscription extends SubscriptionHandler {
...
-      final String rawJson = String.fromCharCodes(base64Decode(message.data!));
+      String rawJson = String.fromCharCodes(base64Decode(message.data!));
+      try {
+        rawJson = String.fromCharCodes(base64Decode(message.data!));
+      } on FormatException {
+        rawJson = message.data!;
+      }
       log.info('rawJson: $rawJson');
       final Map<String, dynamic> json = jsonDecode(rawJson) as Map<String, dynamic>;
       request = BatchRequest.fromJson(json);
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/scheduler/request_subscription.dart:58]: Can we separate the logic into two parts?

```
final String data = message.data!;
Map<String, dynamic> jsonData;
log.info('rawJson: $data');
try {
      jsonData = jsonDecode(data) as Map<String, dynamic>;
    } on FormatException {
      jsonData = json.decode(String.fromCharCodes(base64.decode(data))) as Map<String, dynamic>;
    }

try {
      request = BatchRequest.fromJson(jsonData);
    } catch (e) {
      log.severe('Failed to construct BatchRequest from message');
      log.severe(e);
      throw BadRequestException(e.toString());
    }
```

Comment (yusuf-goog) [reply app_dart/lib/src/request_handlers/scheduler/request_subscription.dart:58]: done

File: app_dart/lib/src/request_handlers/scheduler/request_subscription.dart [hunk 2]
```diff
@@ -46,16 +46,22 @@ class SchedulerRequestSubscription extends SubscriptionHandler {
...
+          jsonData = json.decode(String.fromCharCodes(base64.decode(data))) as Map<String, dynamic>;
+        }
+
+    try {
+          request = BatchRequest.fromJson(jsonData);
+        } catch (e) {
+          log.severe('Failed to construct BatchRequest from message');
+          log.severe(e);
+          throw BadRequestException(e.toString());
+        }
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/scheduler/request_subscription.dart:64]: I suspect a format is needed here to pass the CQ.

**Review by keyonghan — APPROVED** [review]
LGTM with a nit.

File: app_dart/lib/src/request_handlers/scheduler/request_subscription.dart (hunk 2 above)
Comment (yusuf-goog) [reply app_dart/lib/src/request_handlers/scheduler/request_subscription.dart]: done

**Pull request merged**